File: ICPC contests/Brazil Subregional Programming Contest 2022-2023/D.hs
-- https://codeforces.com/gym/103960/problem/D

parse :: String -> (Word, Word, Word)
parse ss =
  let [n, x, y] = map read . words $ ss
   in (n, x, y)

solve1D :: Word -> Word -> Word
solve1D n x
  | x == 2 ^ (n - 1) = 0
  | x < 2 ^ (n - 1) = 1 + solve1D n (2 * x)
  | otherwise = solve1D n (2 ^ n - x)

solve :: (Word, Word, Word) -> Word
solve (n, x, y) = solve1D n x

output :: Word -> String
output = show

main :: IO ()
main = interact $ output . solve . parse

File: ICPC contests/Brazil Subregional Programming Contest 2022-2023/D.rs
// https://codeforces.com/gym/103960/problem/D

fn solve(n: usize, x: usize) -> usize {
    if x == (1 << (n - 1)) { // 2 ^ (n - 1)
        0
    } else if x < (1 << (n - 1)) { // 2 ^ (n - 1)
        1 + solve(n, x * 2)
    } else {
        solve(n, (1 << n) - x)
    }
}

// main function
fn main() {
    // Read 3 naturals from stdin
    let mut line = String::new();
    std::io::stdin().read_line(&mut line).unwrap();
    let mut iter = line.split_whitespace();
    let n: usize = iter.next().unwrap().parse().unwrap();
    let x: usize = iter.next().unwrap().parse().unwrap();
    // No need to parse y

    let res: usize = solve(n, x);
    println!("{res}");
}
